### DISCUSSION TITLE
Teensy 4.1 (ILI9341_t3 + XPT2046_Touchscreen)

### DISCUSSION THREAD

**Discussion by MkMunich** [ImpulseAdventure/GUIslice#369] [General]
Hi,
I found GUIslice and would like to use this great software with my teensy 4.1 board.
Some days ago teensy 3 support had been added to GUIslice, so I thought it might be possible to update to the latest teensy device. I got ILI9341_t3 (// https://github.com/PaulStoffregen/ILI9341_t3) working without GUIslice using PlatformIO. So I created a new PlatformIO project starting from the GUIslice.ino file generated by GUISliceBuilder. I userd a few diffent control elements - just to test them. I struggled a while to set up platformio.ini (as I'm new to PlatformIO). Finally I got compile clean.
I found the instructions how to configure my used pins in Wiki. So I'd put the pins into ./configs/teensy-adagfx-ili9341-xpt2046-audio.h and uncommented the line in GUIslice_config.h to include that header.

BUT there is no #define for the touch's interrupt pin (this ist used in my working // https://github.com/PaulStoffregen/ILI9341_t3 project:
    #define TS_CS   3
    #define TS_IRQ  2  // 255 = unused
    XPT2046_Touchscreen ts = XPT2046_Touchscreen(TS_CS, TS_IRQ);

Is GUIslice not using a touch interrupt, but polling instead?

To understand, how GUIslice's touch events are processed, I would apperetiate getting a hint...
Could I hook the touch events from Paul's ILI9341_t3 into GUIslice?

Best regards
MK_Munich

**Comment by ImpulseAdventure** [comment]
Hi - yes, the library defaults to use polling with XPT2046_Touchscreen at the moment. The config file you are using is the correct one.

Can you confirm that the XPT2046_Touchscreen/TouchTest sketch works if you uncomment the chip-select instantiation? (This should match what GUIslice is using)

https://github.com/PaulStoffregen/XPT2046_Touchscreen/blob/aef8ec703e2a04617a331e127d441b58e549c01a/examples/TouchTest/TouchTest.ino#L9

You should see this line near the top of the TouchTest sketch:
XPT2046_Touchscreen ts(CS_PIN);

thanks

**Comment by MkMunich** [comment]
Hi Calvin,
yes, I can confirm, that the touch IRQ is working in my test PlatformIO project.
Also good news, that I got compile & link clean with PlatformIO including the deb lib impulseadventure/GUIslice@^0.16.0. And also my example is running on my teensy 4.1 device (Touch events are detected, only calibration needs to be adjusted) .
I searched the c'tor call to XPT2046_Touchscreen ts(CS_PIN); in GUIslice and did not yet ind it. According to Paul's lib the second optional parameter (IRQ pin) defaults to not using IRQ.
Last not least I worry, that the correct define for teensy's 4.1 mcu may be not set correctly be the board core. I'll need to figure out, whether I'll need to define it in the plaformIO.ini file myself (e.g. as build flag:	-D __IMXRT1062__)  {HINT: There is a problem in the representation of two underline characters here in this editor: when wiriting & editing the two underlines are seen. After saving the post, they are disappearing - strange. Correct is: under under IMXRT1062 under under !}.
I'll report...
Best Regards

**Comment by MkMunich** [comment]
... The default calibration looks ok. ...I forgot to ask about rotation: In my configuration file './configs/teensy-adagfx-ili9341-xpt2046-audio.h' I've chosen a rotation:   #define GSLC_ROTATE     1
and now I'm not sure, whether|how the touch events are taking this setting into account? A short test showed, that (in Paul's test pgm the lower left corner provides touch values ~[100,100]) in my pgm y-axis is mirrowed. Good news: the touch events are provided to its callback, when I also mirror the touch position myself...

**Reply by ImpulseAdventure** [reply]
Hi @MkMunich -- sorry for the delay as it seems I'm not always notified about discussion thread updates.

First, as you are using the ILI9341_t3 library, I think you may want to use this config file instead:
- `teensy-adagfx-ili9341_t3-xpt2046-audio.h`
- The config file you mention above (teensy-adagfx-ili9341-xpt2046-audio.h) would probably load the Adafruit ILI9341 library instead of Paul Stoffregen's modified T3 library.

From your last update, I understand that the touch is now working. Did you modify GUIslice to use the IRQ constructor? (You would find it in GUIslice_drv_adagfx.cpp near `#elif defined(DRV_TOUCH_XPT2046_PS)`). For example:
https://github.com/ImpulseAdventure/GUIslice/blob/b2be1c15a39206be96077f7c0e56443f0928814b/src/GUIslice_drv_adagfx.cpp#L529

As for the calibration, you should be able to use the GUIslice diag_ard_touch_calib to adjust the calibration settings for update in Section 4B of your config file.

Regarding the orientation, yes, `GSLC_ROTATE` should be supported by the touch & display handlers within GUIslice. I think there have been some enhancements to some of this code in more recent versions of GUIslice, so you may want to try using the [latest from GitHub](https://github.com/ImpulseAdventure/GUIslice/wiki/Installing-GUIslice-on-Arduino#installing-from-github-source) (as I think you're using the library manager version which is still v0.16.0).

Were there any changes that you needed to make to GUIslice or your platformio file to get the Teensy 4.1 working?

thanks!

**Reply by MkMunich** [reply]
Hi Calvin,

I was already using teensy-adagfx-ili9341_t3-xpt2046-audio.h (forgot to update my post above).

I tried to use the current lib instead of v0.16.0 by changing "lib_deps = impulseadventure/GUIslice@^0.16.0 ..." to "https://github.com/ImpulseAdventure/GUIslice.git". And now PlatformIO is using the latest version :-)

I would suggest the following changes in teensy-adagfx-ili9341_t3-xpt2046-audio.h:
  // Touch bus & pinout
  #define XPT2046_CS     3
  #define XPT2046_IRQ    2 // 255: no IRQ used

And in GUIslice_drv_adagfx.cpp:
#elif defined(DRV_TOUCH_XPT2046_PS)
  const char* m_acDrvTouch = "XPT2046_PS(SPI-HW)";
  #if defined(XPT2046_IRQ)
    // Use SPI, with IRQs
    XPT2046_Touchscreen m_touch(XPT2046_CS, XPT2046_IRQ); // Chip Select pin, IRQ pin
  #else
    // Use SPI, no IRQs
    XPT2046_Touchscreen m_touch(XPT2046_CS); // Chip Select pin
  #endif
  #define DRV_TOUCH_INSTANCE

This would mean to be backward compatible in case the define XPT2046_IRQ is missing &| Paul's new c'tor with 2nd optional parameter IRQ pin is not present.

I'll attach my platformio.ini here (renaming the 
[platformio.txt](https://github.com/ImpulseAdventure/GUIslice/files/6378010/platformio.txt)
extension to txt) - kwnowing, that under construction...

One new question: I tried to get rid of "Adafruit GFX Library" - also thought, that not necessary. Unfortunatelly GUIslice Builder generates the header GUIslice_GSLC.h with #include <Adafruit_GFX.h>. How to deal with this?

**Comment by ImpulseAdventure** [comment]
Excellent — thank you for the followup.

I will add the XPT2046_PS IRQ option as you have suggested.

Regarding the Adafruit_GFX header from the Builder: can you confirm that you have selected the “Teensy” platform / target in the Builder project options before code generation?

thanks

**Reply by MkMunich** [reply]
Good evening Calvin,
you ment this setting?
![2021-04-26 21_13_05-Edit Preferences](https://user-images.githubusercontent.com/26894654/116139331-33f77680-a6d6-11eb-894b-42822192eb82.png)

This is from the generated header file GUIslice_GSLC.h at line ~37:
// ------------------------------------------------
// Headers and Defines for fonts
// Note that font files are located within the Adafruit-GFX library folder:
// ------------------------------------------------
//<Fonts !Start!>
#if defined(DRV_DISP_TFT_ESPI)
  #error Project tab->Target Platform should be tft_espi
#endif
#include <Adafruit_GFX.h>
//<Fonts !End!>



Could it be, that I added the following into GUIslice_config.h ?
#elif defined(__MK66FX1M0__) // Teensy 3.6
  #define GSLC_CFG_ARD
  #define GSLC_DEV_TEENSY
  #define GSLC_DEV_TEENSY_3_6
#elif defined(__IMXRT1062__) // Teensy 4.0   HINT: the define has leading&ending double undelines!
  //#elif defined(ARDUINO_TEENSY40)
  //#elif defined(ARDUINO_TEENSY41)
  #define GSLC_CFG_ARD
  #define GSLC_DEV_TEENSY
  #define GSLC_DEV_TEENSY_4_0
#else
  #warning Unknown device platform
#endif


Best Regards Marcel

**Comment by MkMunich** [comment]
Hi Calvin,
it took a while, but I got my test-GUIslice application running again. My main problem was, that I still have to learn how PlatformIO is working - a great tool and I would never go back to Arduino again...
My problem has been, that I did not find, which config file was taken. Finally I edited all ..\configs\teensy-adagfx-ili9341_t3-xpt2046-audio.h files and configered the used pins in each header file. It should be posible to simply add an include path in my project\platfromio.ini file, where the order tells PlatformIO to take my own header (a copy in my project\configs directory? - still investigating....

Now, I'm back at the point of looking at the touch events. As mentioned, either my x or y axis is mirrowed. Means instead of the x-value it would be (maxX - x) - value in case x is mirrowed respectively (maxY - y) - value in case the y-axis is mirrowed. I saw flags in GUIslice controlled by the config header file's #define GSLC_ROTATE 0..3. I tried all values and all failed. I wonder, how it is working, as rotation rotates the display AND touch events at the same time. In my case, the touch values are (either x or y) flipped. So it might work, if rotation is separately set for display and touch. The main problem is, that my touch events do not need  a 'rotation', but a flip.
Is this already possible in GUIslice?

Best Regards
Marcel

**Reply by ImpulseAdventure** [reply]
Hi - I can followup on your config / Builder updates later, but for the moment I’d like to address your touch flip:

When you run the diag_ard_touch_calib, any touch rotation / flip differences detected versus your screen are supposed to be factored into your touch config calibration settings.

When you ran the diag_ard_touch_calib sketch, the last part of the test (when it lets you draw red dots on the screen) should show you how the calibrated touch settings look. In this part, did the red dots track your touches or were they also flipped?

thanks

**Reply by MkMunich** [reply]
Hi, I got the calibration running (needed to comment one line to get compile clean).
in GUIslice_GLSC.h line 62:
//enum {E_BUILTIN5X8, MAX_FONT};

There was/is no display-element seen to touch at during calibration... I just pressed each corner clock-wise starting from bottom right for more than 3 secs.
The log is as below:
--- Available filters and text transformations: colorize, debug, default, direct, hexlify, log2file, nocontrol, printable, send_on_enter, time
--- More details at http://bit.ly/pio-monitor-filters
--- Miniterm on COM6  9600,8,N,1 ---
--- Quit: Ctrl+C | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
Capture TL: X=229 Y=480 Cnt=1223, Z=(1851..2602)
Capture TR: X=3814 Y=509 Cnt=1312, Z=(1744..2079)
Capture BL: X=3783 Y=3807 Cnt=717, Z=(401..442)
Capture BR: X=209 Y=3806 Cnt=995, Z=(626..1751)
SegDetect: Unknown display orientation (0,2,8,6)

ERROR: Calibration failed. Please check touch pin config ADATOUCH_PIN_*

CALIB: Result(adj) XMin=2005 XMax=2011 YMin=383 YMax=3916

CALIB: Rotate
Capture TL (Rotated): X=2384 Y=2353 Cnt=23, Z=(951..1554)
ERROR: Could not detect rotation. Please retry calibration.


Recommended calibration settings are shown below.
- Copy and paste these lines into your GUIslice config file
  over top of the existing ADATOUCH_X/Y_MIN/MAX settings.

---
  // DRV_TOUCH_XPT2046_PS [240x320]:
*** ERROR: Calibration failed. Please check log.

**Comment by Pconti31** [comment]
@MkMunich
`Good evening Calvin,
you ment this setting?`
Ah, No.  You do have teensy set correctly in general settings so that when you create a new project it will default to target teensy, but you need to double check your test project is set for teensy.  You do this by opening your test project in the builder and clicking on the Project Options tab and it must show teensy for your target as so:
![profile](https://user-images.githubusercontent.com/44075106/116297272-18897b80-a769-11eb-88e2-008630efc4c7.png)
Paul--

**Reply by MkMunich** [reply]
Thx Paul, I got it and was able to get rid of 'Adafruit GFX Library'. Best Regards Marcel

**Comment by ImpulseAdventure** [comment]
Mi Marcel --

Glad to hear that things seem like they are working. I assume the calibration has addressed your touch flip.

I have now made an update to the latest GUIslice to support the `XPT2046_IRQ` setting for the XPT2046_PS  library. The associated config files have also been updated.

To wrap things up, I thought I would see if there were any remaining changes that we needed to make based on your findings?
- In https://github.com/ImpulseAdventure/GUIslice/discussions/369#discussioncomment-661456 you mentioned a change to GUIslice_config.h for the Teensy 4.1. There, it looked like you might have used `defined(IMXRT1062)` instead of what I had, which was `defined(__IMXRT1062__)`. Can you confirm that this was necessary?

thanks

**Reply by MkMunich** [reply]
Hi Calvin,
sorry about the mess... double under chars are not correctly displayed in all posts.
#elif defined(under under IMXRT1062 under under) // Teensy 4.0
But I was not sure, whether I had inserted it. I can confirm, that it is now this line:
#elif defined(__ IMXRT1062 __) // Teensy 4.0 (HINT: the leading an ending double under chars are present in my code).
In my actual platformio.ini file I'm using the following:
[env:teensy41]
platform = teensy
board = teensy41
framework = arduino
build_flags = 
	-D __IMXRT1062__ (also with leading and ending under under chars)

I searched for "IMXRT1062" without any under :-) in all files and checked, that always occuring with leading&ending under under chars). So, this looks good.
It compiles and I can see the example GUI, I created in GUIslice Builder on my device. I also get callback events, but still having the issue, that I need to flip the touch position myself (SW should do this much better than me - I'm just guessing the position of the GUI elements :-).

Did you see my calibration values (& error message of the calibration app)? I did not see a log-file and I think, the calibration app did not provide proper output to get copied into the header file GUIslice_config.h as I would expect lines starting with: #define - correct?

Best Regards
Marcel

**Reply by MkMunich** [reply]
Hi, meanwhile I figured out, that there is a #define ADATOUCH_REMAP_YX with (in principle) values from 0..3. I tried all values (and saw lateron, that 1 test should have been enough, as bTouchRemap is just bool :-)
ADATOUCH_REMAP_YX does not fix my problem as it just swap X and Y Touch values. Remind that I'll need flip of either X or Y. Mathematically something like:  Xresult = XMax - Xtouch.

**Reply by MkMunich** [reply]
OK, I got it. After digging into the code I had the idea to swap the ADATOUCH_X_MAX and ADATOUCH_X_MIN values. This is indeed the easiest solution of my problem.

Many thanks to Calvin and Paul for your help. Let me know, if you would like to get my final platformio.ini file.
Best Reagrds
Marcel

**Reply by ImpulseAdventure** [reply]
Hi Marcel -- yes, swapping the `ADATOUCH_X_MAX/MIN` values will effectively flip the horizontal calibration.
However, the calibration sketch should be able to provide you with these values.

In https://github.com/ImpulseAdventure/GUIslice/discussions/369#discussioncomment-666007 above, you indicated that nothing was shown on the screen when you ran the `diag_ard_touch_calib` sketch... can you confirm that for me? Normally, you should see some textual instructions and then a series of crosshairs to touch on. If you are not seeing this, then the calibration will likely generate an error message like the one you encountered.

thx

**Reply by MkMunich** [reply]
Hi Calvin, may be I forgot to 'Clean' before rebuilding the calibration. I checked, that using the latest sources (may be, this helped, too). So, it's all working a expected.

Just a short question... Should I see any element when asked to "Confirm red dots match touches" ?

For interested people:
Here a description, how I got my teensy 4.1 GUIslice test application working:
First I created a new directory e.g. 'TestGUI' in the PlatformIO projects folder. Within I created a subdirectory 'scr'. Then I created a demo GUI using GUIslice Builder. I configured 'Target platform' to be 'teensy' (in Edit->Options tab General). Double check, that 'teensy is set in "E_PROJECT_OPTIONS" property 'Target platform' as well. Now, I saved this GUIslice project into "TestGUI\src". Finally I generated the code and closed GUIslice Builder. In the next step I prepared the PlatformIO project by creating a platformio.ini file in "TestGUI". The one here is a good starting point (rename extension .txt to .ini).
[platformio.txt](https://github.com/ImpulseAdventure/GUIslice/files/6409411/platformio.txt)

Let's have a look at some details: I struggled a lot which 'teensy-adagfx-ili9341_t3*' config file I needed and how to configure PlatformIO, that exactly my config file is taken (and no other of many copies). Luckily, Calvin prepared the sources, so the config file can be given from outside via #define (refer to section [common] build_flags). For this, the config file (in my case a copy of 'teensy-adagfx-ili9341_t3-xpt2046-audio.h') must be present in directory 'TestGUI\include'. If you take an other config file, then change section "[common] build_flags = " in platformio.ini accordingly.

Important step: adjust the config file according to your wiring. In my case I needed to activate the audio shield SD by "#define GSLC_SD_EN    1"
Section [env:Teensy-ILI9341_t3-XPT2046] is the one, which is used for build, upload, clean in the PlatformIO GUI. Currently, I'm using the actual verion of GUIslice library. If you would like to use a released version (e.g. 0.16.0) then replace in section "[env:Teensy-ILI9341_t3-XPT2046] lib_deps" line "https://github.com/ImpulseAdventure/GUIslice.git" by "impulseadventure/GUIslice@^0.16.0".

I like to have a "TestGUI.code-workspace" in 'TestGUI' allowing to open PlatformIO by a double click on this file. The content is similar to the following:
{
	"folders": [
		{
			"name": "TestGUI",
			"path": "."
		}
	],
	"settings": {
	}
}

Ready for the first PlatformIO build of environment "Teensy-ILI9341_t3-XPT2046" ? It should compile, link and can be uploaded to teensy 4.1. If your display is correctly wired, then you should see your GUI example. At this point the programming of your GUI can start.

If you like to calibrate your touch display, then you can do this as follows:
Copy the file "diag_ard_touch_calib.ino" into "TestGUI\src". Then rename "TestGUI.ino" to "TestGUI.ino_" (make it unusable for PlatformIO compiler). Then compile and upload as normal and follow the instructions in the touch screen. At the end rename "TestGUI.ino_" back to "TestGUI.ino" and rename "diag_ard_touch_calib.ino" to "diag_ard_touch_calib.ino_" to make this calibration app not compilable.

Enjoy GUI programming.

**Reply by ImpulseAdventure** [reply]
Excellent!

Thank you very much for sharing your platformio configuration and steps used. This will be very helpful for others. I should probably create a wiki page for this and include some of your steps there.

As for the calibration — the “red dots” phase of the calibration simply draws red dots where touches are detected (taking into account the latest calibration settings). The display orientation will rotate rotating after every release. No GUI elements will be shown during this test.
